src/App.jsx

```diff
@@ -108,35 +73,83 @@ function App () {
         window.lodash = _.noConflict()
         setPsiturk(new PsiTurk(turkUniqueId, '/complete'))
         setMethod('mturk')
-        setLoggedIn(true, 'mturk', turkUniqueId)
+        // TODO: This won't work
+        handleLogin('mturk', turkUniqueId)
         /* eslint-enable */
       } else if (config.USE_PROLIFIC) {
         const pID = getProlificId()
         if (config.USE_FIREBASE && pID) {
           setMethod('firebase')
-          setLoggedIn(true, 'prolific', pID)
+          // TODO: This wont work
+          handleLogin('prolific', pID)
         } else {
-          setReject(true)
+          // Error - Prolific must be used with Firebase
+          setIsError(true)
         }
-      } else if (config.USE_FIREBASE) {
-        setMethod('firebase')
-        // Autologin with query parameters
+      } else if (config.USE_FIREBASE) { 
+        // Fill in login fields based on query parameters (may still be blank)
+        const query = new URLSearchParams(window.location.search)
         const participantId = query.get('participantID')
         const studyId = query.get('studyID')
-        if (participantId) {
-          setEnvParticipantId(participantId)
-        }
-        if (studyId) {
-          setEnvStudyId(studyId)
-        }
+        if (participantId) setParticipantID(participantId)
+        if (studyId) setStudyID(studyId)
+
+        setMethod('firebase')
       } else {
         setMethod('default')
       }
     }
+    console.log(currentMethod)
     // eslint-disable-next-line
   }, [])
 
-  if (reject) {
+  
+
+  /** VALIDATION FUNCTIONS */
+
+  // Default to valid
+  const defaultValidation = async () => true
+
+  // TODO: This should JUST validate, not init the participant
+  // Validate participant/study against firestore rules
+  const firebaseValidation = (participantId, studyId) => {
+    // return initParticipant(participantId, studyId, startDate)
+    return validateParticipant(participantId, studyId)
+  }
+
+  /** DATA WRITE FUNCTIONS */
+
+  // Adding data functions for firebase, electron adn Mturk
+  const defaultFunction = () => {}
+  const firebaseUpdateFunction = (data) => { addToFirebase(data) }
+  const desktopUpdateFunction = (data) => { ipcRenderer.send('data', data) }
+  const psiturkUpdateFunction = (data) => { psiturk.recordTrialData(data) }
+
+  /** EXPERIMENT FINISH FUNCTIONS */
+
+  const defaultFinishFunction = (data) => { data.localSave('csv', 'neuro-task.csv') }
+  const desktopFinishFunction = () => { ipcRenderer.send('end', 'true') }
+  const psiturkFinishFunction = () => {
+    const completePsiturk = async () => {
+      psiturk.saveData()
+      await sleep(5000)
```

Comment:
Rather than this sleep statement, looks like there are callbacks.

[Source](https://psiturk.readthedocs.io/en/latest/api.html#psiturk-savedata-callbacks)

```
psiturk.saveData({
   success: function() {
      // function to run if the data is saved
   },
   error: function() {
      // function to run if there was an error
   }
});
```

Reply:
Update to use the callbacks instead of sleeping

Reply:
Somewhat unrelated but this seems to be a python package? I'm not entirely sure how we're executing the code in our project. It looks like it's passed as a script in `index.html` but why not just include it as a package?

Reply:
I did refactor to use the callbacks